<img src="http://imgur.com/1ZcRyrc.png" style="float: left; margin: 20px; height: 55px"> 

## Lab: SQL Practice

_Author: Matt Brems_

### Required Preparation

In order to run SQL queries within Python/this Jupyter notebook, you will need to install the following:

```bash
conda install sqlalchemy
conda install psycopg2
```

If you get an error trying to install `psycopg2`, you can install a stand-alone package (sufficient for everything you'll need to complete this lab) by entering the following anywhere in the terminal: 

```bash 
pip install psycopg2-binary
```

(If `pip` doesn't work for you, try `pip3 install psycopg2-binary`.)

### Instructions

In this lab, you're going to have ten prompts. Each prompt will require you to do two things:
1. Write a SQL query.
2. Use the result of that query to answer the question.

The purpose of this lab is to get you to practice your SQL - an **important** skill in data science! While it is possible to get these answers using Pandas, it should be your goal to do everything only in SQL.

**Data**: The data used in this lab is the Iowa liquor database. It contains various tables of transactions. The schema for this database is here:

<img src="./images/schema.png" alt="schema" width="750"/>

In [1]:
# Import Pandas and Create_Engine


from sqlalchemy import create_engine
import pandas as pd

# Set up the engine to access the data.
engine = create_engine('postgresql://analytics_student:analyticsga@analyticsga-psql.generalassemb.ly:5432/iowa_liquor_sales_database')

### Prompt 1

How many rows are there in the `stores` table of the Iowa liquor database?

In [2]:
# Use a SQL query to find the answer to the above prompt.

sql = """

SELECT COUNT(*) AS total_rows
FROM stores

"""


pd.read_sql_query(sql, engine)


total_rows
0        1973

**Prompt 1 Answer**: 1973

### Prompt 2

If you sort the stores in alphabetical order by name, which store is fourth?

In [3]:
# Use a SQL query to find the answer to the above prompt.

sql = """

SELECT store, name
FROM stores
ORDER BY name
LIMIT 4;

"""

pd.read_sql_query(sql, engine)

store                                name
0   4986  218 Fuel Express & Chubby's Liquor
1   3047          3047 Cub Foods / Iowa City
2   3060               3060 Cub Foods / Ames
3   3061         3061 Cub Foods / Sioux City

**Prompt 2 Answer**: 3061

### Prompt 3

Among those sales whose category name is `IMPORTED VODKA`, if you sort by vendor in alphabetical order, what is the bottle price and number of bottles bought in the first 3 transactions?
> You should give us three sets of numbers - one for each transaction.

In [4]:
# Use a SQL query to find the answer to the above prompt.

sql = """

SELECT btl_price, bottle_qty, vendor, category_name
FROM sales
WHERE LOWER(category_name) LIKE '%%imported vodka%%'
ORDER BY vendor;
"""

pd.read_sql_query(sql, engine)

btl_price  bottle_qty                     vendor   category_name
0          $8.99          12  Adamba Imports Int'l Inc.  IMPORTED VODKA
1         $28.50           6      A Hardy / U.S.A. Ltd.  IMPORTED VODKA
2         $28.50          30      A Hardy / U.S.A. Ltd.  IMPORTED VODKA
3         $28.50          30      A Hardy / U.S.A. Ltd.  IMPORTED VODKA
4         $28.50           3      A Hardy / U.S.A. Ltd.  IMPORTED VODKA
...          ...         ...                        ...             ...
205380    $16.33           2                       None  IMPORTED VODKA
205381    $16.33           3                       None  IMPORTED VODKA
205382    $16.33           2                       None  IMPORTED VODKA
205383    $16.33           2                       None  IMPORTED VODKA
205384    $16.33           1                       None  IMPORTED VODKA

[205385 rows x 4 columns]

**Prompt 3 Answer**: 205385


### Prompt 4

What is the total amount of liquor sold in February 2015?

In [5]:
# Use a SQL query to find the answer to the above prompt.
import datetime

sql = """
select SUM(total) AS sales_in_feb2015
FROM sales
WHERE text(date) LIKE '2015-02%%';
"""
pd.read_sql_query(sql, engine)

sales_in_feb2015
0       21295350.55

**Prompt 4 Answer**: 21295350.55

### Prompt 5

Among all transactions where `IMPORTED VODKA` was sold, which vendor has the most transactions? Report the total number of transactions, the number of bottles sold, and the total amount of revenue from these transactions. Rename the columns to make sure there isn't any confusion as to what they mean!

In [6]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT vendor, total AS total_revenue, bottle_qty AS total_bottles_sold
FROM sales
WHERE category_name LIKE '%%IMPORTED VODKA'
GROUP BY vendor, total, bottle_qty
ORDER BY bottle_qty DESC
LIMIT 5;

"""

pd.read_sql_query(sql, engine)

vendor  total_revenue  total_bottles_sold
0   Constellation Wine Company Inc.        34522.8                1560
1                   Jim Beam Brands        27180.0                1500
2  Pernod Ricard USA/Austin Nichols        29686.8                1320
3                   Jim Beam Brands        21744.0                1200
4  Pernod Ricard USA/Austin Nichols        35988.0                1200

**Prompt 5 Answer**: Constellation Wine Company Inc. has most transactions and Pernod Ricard USA/Austin Nichols has highest revenue

### Prompt 6

The reason we sell things is to make money. In sales, the term **markup** means the amount of extra money charged for a product over the cost to make that product. (You can think of the markup as the profit for that product.)

In Iowa, the law states the minimum price at which liquor may be sold.

Calculate the markup (name this `Markup`) that shows the difference between the `btl_price` and the `state_btl_cost`. 

> If you haven't subtracted columns together before, [this StackOverflow question](https://stackoverflow.com/questions/7536996/multiplying-two-columns-in-sql-server/7537059) may be helpful.

In [7]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT btl_price, state_btl_cost, (btl_price - state_btl_cost) AS Markup
FROM sales
LIMIT 5;

"""

pd.read_sql_query(sql, engine)

btl_price state_btl_cost markup
0    $17.24         $11.49  $5.75
1    $17.24         $11.49  $5.75
2    $17.24         $11.49  $5.75
3    $25.73         $17.15  $8.58
4     $7.50          $5.00  $2.50

### Prompt 7

Calculate the "Percentage Markup" by dividing markup by the state bottle cost. (Rather than the total profit per bottle, this will tell you how much stores are increasing their price over the state minimum.)

After calculating the "Percentage Markup," calculate the average percentage markup per **store**. Which store has the largest average markup, and what is that average markup?

In [8]:
sql = """
SELECT store, AVG((btl_price - state_btl_cost) / state_btl_cost) AS "Avg Percentage Markup"
FROM sales
GROUP BY store
ORDER BY "Avg Percentage Markup" DESC
LIMIT 5;
"""

pd.read_sql_query(sql, engine)

store  Avg Percentage Markup
0   4013               0.525471
1   4024               0.521499
2   4266               0.518903
3   3833               0.515094
4   3660               0.514591

**Prompt 7 Answer**:  store : 4013 with an average markup percentage of 0.525

### Prompt 8

Building off of your last query, retrieve the five stores with the highest average markup percentage. Along with the store IDs and average markup percentage, please return the minimum and maximum markup percentage for those stores, as well as the names and addresses of those store.

What are the names of the five stores?

In [9]:
sql = """
SELECT sales.store, AVG((btl_price - state_btl_cost) / state_btl_cost) AS "Avg Percentage Markup", stores.name,\
stores.store_address, MIN(((btl_price - state_btl_cost) / state_btl_cost)) AS "MIN_PCT", \
MAX((btl_price - state_btl_cost) / state_btl_cost) AS "MAX_PCT"
FROM sales INNER JOIN stores ON stores.store = sales.store
GROUP BY sales.store, stores.store
ORDER BY "Avg Percentage Markup" DESC
LIMIT 5;
"""

pd.read_sql_query(sql, engine)

store  Avg Percentage Markup                        name  \
0   4013               0.525471      Wal-Mart 0841 / Tipton   
1   4024               0.521499  Wal-Mart 1546 / Iowa Falls   
2   4266               0.518903  Wal-Mart 1683 / Shenandoah   
3   3833               0.515094    Wal-Mart 3394 / Atlantic   
4   3660               0.514591   Wal-Mart 2935 / Knoxville   

                                       store_address   MIN_PCT   MAX_PCT  
0  1126 Highway 38 North\nTipton, IA 527720000\n(...  0.498571  6.708763  
1  840 S Oak\nIowa Falls, IA 501260000\n(42.50295...  0.498462  6.708763  
2  705 S Fremont\nShenandoah, IA 516010000\n(40.7...  0.498462  6.708763  
3  1905 East 7th St\nAtlantic, IA 500220000\n(41....  0.498462  6.708763  
4  814 W Bell Ave\nKnoxville, IA 501380000\n(41.3...  0.498462  6.708763

**Prompt 8 Answer**: Wal-Mart 0841/Tipton, Wal-Mart 1546 Iowa Falls, Wal-Mart 1683 Shenandoah, Wal-Mart 3394 / Atlantic, Wal-Mart 2935 / Knoxville

### Prompt 9

Which nine counties sell the highest number of bottles of liquor per capita (per person)? Calculate the average bottle size of the bottles of liquor sold by these counties rounded to the nearest first decimal place. Of the nine counties selling the highest number of bottles of liquor per capita, which county sells (on average) the largest bottle size?

**_BONUS FUN_**: This question can (also) be solved with a SQL subquery - that is, a query that queries a query! If you'd like to give this a try, you can read more about it [here](https://www.dofactory.com/sql/subquery). Otherwise, keep an eye out for the solution key!

In [18]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT county, ROUND(AVG(a.bottle_size),1) AS "avg_bottle_size",  bottle_qty
FROM(SELECT  s.county, s.bottle_qty, s.total, p.bottle_size
FROM sales AS s INNER JOIN products as p ON s.item = p.item_no
GROUP BY s.county, s.bottle_qty, s.total, p.bottle_size
ORDER BY s.total DESC
LIMIT 9)a
GROUP BY county, bottle_qty
ORDER BY avg_bottle_size DESC
LIMIT 9;
"""

pd.read_sql_query(sql, engine)

county  avg_bottle_size  bottle_qty
0         Dallas           1750.0        1800
1         Dallas           1750.0        2400
2  Pottawattamie           1750.0        3000
3        Johnson           1000.0        1680
4           Polk           1000.0        1728
5           Polk           1000.0        2400
6           Polk           1000.0        3456
7           Polk            750.0        2148
8         Dallas            750.0        2700

**Prompt 9 Answer**:

### Prompt 10 (advanced but required!)

In the `sales` table, three columns are `btl_price`, `bottle_qty`, and `total`. The `total` for a transaction _should be_ the product of `btl_price` and `bottle_qty`. How many transactions have a value of `total` that is not equal to `btl_price` time `bottle_qty`?

> There will be a type error as well! We cannot compare type `money` to type `real`. We have to convert `total` to type `money` so that we can compare the two directly. [This link](https://docs.microsoft.com/en-us/sql/t-sql/functions/cast-and-convert-transact-sql?view=sql-server-ver15#syntax) may be helpful to you!

In [52]:
# Use a SQL query to find the answer to the above prompt.

sql = """

SELECT btl_price, bottle_qty, CAST(total as money) AS "total1", btl_price * bottle_qty AS "new_total"
FROM sales
WHERE CAST(total as money) != (btl_price * bottle_qty);

"""

pd.read_sql_query(sql, engine)

Empty DataFrame
Columns: [btl_price, bottle_qty, total1, new_total]
Index: []

**Prompt 10 Answer**: None